util.py

```diff
@@ -57,6 +62,37 @@ def build_menu(buttons,
     return menu
 
 
+def rate_limit_tracker(update: Update, context: CallbackContext):
+    data = context.chat_data.get('rate_limit', {})
+
+    for key in data.keys():
+        data[key] += 1
+
+
+def rate_limit(f):
+    """
+    Rate limit command so that RATE_LIMIT_SPACING non-command messages are
+    required between invocations.
+    """
+    @wraps(f)
+    def wrapper(update, context, *args, **kwargs):
+        # Get rate limit data
+        try:
+            data = context.chat_data['rate_limit']
+        except KeyError:
+            data = context.chat_data['rate_limit'] = {}
+
+        # If we have not seen two non-command messages since last of type `f.__name__`
+        if data.get(f.__name__, RATE_LIMIT_SPACING) < RATE_LIMIT_SPACING:
```

Comment:
Pretty sure functions are hashable. I think it's better to use the function as the key rather than the name which can be duplicated. Better, do `f._field` so the data is assigned to the function.

Reply:
Ah yeah I do believe you're right.
I'd rather not assign the data directly on the function tho, since doing that would men that it's the same rate-limit applies to both the offtopic group on the ontopic one. It would also make it much harder to iterate over it in `rate_limit_tracker`

Reply:
Functions could have a dict with per-chat data, but yeah, no need to go that way. Just use `f` as the index should be good.